Implement Results Card Design

Icemanbeta opened the pull request:
- Created reusable layout components `Row`, `Column` for faster templating
- `ResultsCardDataItem` => `DefinitionList` (`components/DefinitionList`)
- `ResultsCardDataPoint` => `Definition` (`components/DefinitionList/Definition`)
- Depreciated `ResultsCardDataSection`
- Update `Results` mock object to the latest from **State-TalentMAP-API`**
- Created reusable definition wrappers: `DefinitionList`, `Definition`
- Created unit tests for all component use cases
- Added `_lists.scss` and `@import` to scss entry file
- Synced latest design guide wireframe palette (with names) to `variables.scss` and cleaned up duplicate styles by referencing the source of truth.
- Cleaned up `variables.scss` and added section comment blocks so it's easier to read
- Added import of `variables.scss` from `uswds` and their bourbon functions so we can extend/implement the core if we need to
<img src="https://user-images.githubusercontent.com/1786875/37638656-6a9cfedc-2be4-11e8-8717-894c832b313f.png" alt="Color" width="400">
- Updated `@import` paths to be module routed instead of relative  in`styles.scss`
- Added `truncated` prop to `Definition` for truncating text overflow via ellipsis
- Updated `InteractiveElement` to allow for more than `div` and `span` due to `usa-button` unfavorable overriding needs
- Added automatic disabling and active states via classes and attributes on `CompareCheck` and `className` proxy prop
- Implemented new design to `ResultsCard`
- Updated `SystemMessages.js` for easier maintenance (alphabetical)

mjoyce91 commented on src/sass/_overrides.scss:
```diff
@@ -1,3 +1,7 @@
+.usa-button {
+  font-family: 'Roboto';
```
This threw off the Search button on the homepage. If we drop the font-size of that button from `.9em` to `.8em`, it sits correctly. See below:

This PR:
<img width="368" alt="screen shot 2018-03-20 at 4 03 03 pm" src="https://user-images.githubusercontent.com/11576347/37679742-386af80a-2c58-11e8-848a-873ba36a47ee.png">

Before:
<img width="357" alt="screen shot 2018-03-20 at 4 03 11 pm" src="https://user-images.githubusercontent.com/11576347/37679751-441a2572-2c58-11e8-8e3e-03ffa9b19e16.png">

Changes to `.8em`:
<img width="258" alt="screen shot 2018-03-20 at 4 02 54 pm" src="https://user-images.githubusercontent.com/11576347/37679770-4dc2e974-2c58-11e8-8bcc-6d344b7cf95e.png">

mjoyce91 replied:
Also throws off the View Post Data button on the Compare page

<img width="175" alt="screen shot 2018-03-21 at 8 49 54 am" src="https://user-images.githubusercontent.com/11576347/37710728-e1dc3232-2ce4-11e8-85b0-336a46ce970c.png">

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -1,78 +1,119 @@
+
+  options.compare = {
+    as: 'button',
+    refKey: result.id,
```
We use `position_number` for the compare array since that's what the query on the `/compare` route uses.

Icemanbeta replied:
It's not unique

mjoyce91 replied:
The way the data set is generated is weird right now, but within State it will be unique.

mjoyce91 commented on src/Components/CompareCheck/CompareCheck.jsx:
```diff
@@ -56,29 +56,54 @@ class CompareCheck extends Component {
-    const iconClass = this.getSavedState() ? 'check-square-o' : 'square-o';
+    const props = this.props;
+    const isChecked = this.getSavedState();
+    const text = this.isDisabled() ? 'Limit Reached' : 'Compare';
```
I like the style used here! Unfortunately the color contrast is too low:

<img width="914" alt="screen shot 2018-03-20 at 4 10 33 pm" src="https://user-images.githubusercontent.com/11576347/37680114-51823d84-2c59-11e8-8dea-5b4107196824.png">

mjoyce91 replied:
I think we can also bump up the font-size on the buttons to match the content of the card.

Icemanbeta replied:
If I bump the font size, the button wraps with the loading icon in the favorites button.

What should I brighten or dim for contrast? This is a disabled state though and it's standard for uswd and well we use the same button palette apparently for colors

mjoyce91 replied:
Ah okay. This should be good then.

And that's a good point, we should actually be good on color contrast then.

Disregard both comments! :+1:

Icemanbeta replied:
Fixed! e2a7fa1ca6e38abee708f5b7569260a0029694ec

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -1,78 +1,119 @@
+      'Grade': _.get(result, 'grade', ''),
+      'Post': _.get(result, 'post.location', ''),
+      'Skill Code': _.get(result, 'skill', ''),
+      'Language': _.get(result, 'languages[0].language', ''),
```
I would look into our `SystemMessages` file and our `propOrDefault` function. We have some default text that we display incase some of these values are null. For example, in our Details page:

<img width="474" alt="screen shot 2018-03-20 at 4 12 41 pm" src="https://user-images.githubusercontent.com/11576347/37680206-925a6ef8-2c59-11e8-8e21-2c06df876d27.png">

Opposed to being blank:
<img width="819" alt="screen shot 2018-03-20 at 4 13 15 pm" src="https://user-images.githubusercontent.com/11576347/37680217-9e2b691c-2c59-11e8-8400-1b96ec46d02d.png">

Icemanbeta replied:
So `get` does the same thing but it works for collections and arrays. Actually in the interceptor PR I updated the `propOrDefault` to point at `get` from lodash.

mjoyce91 replied:
Ah okay. Even without using `propOrDefault` here, we can pass in the related `SystemMessage` as the third/`default` argument.

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -1,78 +1,119 @@
+          <DefinitionList items={sections[0]} />
+        </Column>
+      </Row>
+      <Row id={result.id} fluid>
```
We don't need to be perfectly responsive right now, but let's fix this:

<img width="845" alt="screen shot 2018-03-20 at 4 17 21 pm" src="https://user-images.githubusercontent.com/11576347/37680425-3b641e72-2c5a-11e8-9645-304201e09e40.png">

mjoyce91 replied:
That's at around 1080px window width

Icemanbeta replied:
Hmm let me add a media query to this. What's the max window width we should support on this app? All the way to mobile?

Icemanbeta replied:
@mjoyce91 `MediaQuery` is pretty awesome. It has everything you can ever need 👍 

<img src="https://user-images.githubusercontent.com/1786875/37749929-ab859b84-2d60-11e8-980c-4782326bfd8a.png" height="220" />

The `enforceShort` triggers anytime below `screen-md-max`

9b4485bbffe0e8335c7aaeb6b12800fae56a883c

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -1,78 +1,119 @@
+    <div id={result.id} className="results-card">
+      <Row className="header" fluid>
+        <Column columns="5">
+          <h1>{result.title}</h1>
```
I think this should be `h3`

mjoyce91 commented on src/Components/DefinitionList/DefinitionList.jsx:
```diff
@@ -0,0 +1,69 @@
+        props.children.length ?
+          props.children :
+          items.map(item => (
+            <Definition id="hi" key={shortid.generate()} {...item} />
```
`id="hi"`? :smile:

Icemanbeta replied:
LOL yea totally testing this work around with children fragments in our react version

mjoyce91 commented on src/sass/_variables.scss:
```diff
@@ -1,54 +1,139 @@
+ * Color Palette
+ */
+// Primary
+$primary-blue: #0071BC;
```
check out your `scss-list` config, it'll want all your hex values to be lowercase (`#0071BC` -> `#0071bc`)

Icemanbeta replied:
What wants it lowercase? I actually like it uppercase :cry:

mjoyce91 replied:
Same here, that was just a default from create-react-app. We could change the rule, as long as we're consistent. :thinking:

Icemanbeta replied:
am I supposed to get eslint errors somewhere? lol cause I totally don't get them

mjoyce91 replied:
I use Atom with the `linter-eslint` and `linter-scss-lint` packages installed.

mjoyce91 commented on src/sass/_results.scss:
```diff
@@ -31,57 +30,152 @@
+
+      &:hover {
+        color: $primary-blue-darker;
+        box-shadow:: $button-box-shadow $primary-blue-darker;
```
extra colon here?

Icemanbeta replied:
Sass loader didn't seem to complain weird.

mjoyce91 commented on src/Components/Layout/Column/Column.jsx:
```diff
@@ -0,0 +1,60 @@
+  12: 'one-whole',
+};
+
+const Column = (props) => {
```
I'm excited to use this elsewhere :+1:

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -1,78 +1,119 @@
+    userProfileFavoritePositionHasErrored,
+  } = props;
+
+  const bidCycle = _.chain(result)
```
Maybe a comment here on what this does

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -39,20 +68,20 @@ const ResultsCard = (props) => {
+      'Grade': getResult('grade', NO_GRADE),
+      'Post': getResult('post.location', NO_POST),
+      'Skill Code': getResult('skill', NO_SKILL),
+      'Language': getResult('languages[0].language', NO_LANGUAGES),
```
I forgot to mention - we have a `LanguageList` component to handle languages. This is because in some instances, there are multiple languages. The component will format them nicely for you.

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -39,20 +68,20 @@ const ResultsCard = (props) => {
-      'Incumbent': _.get(result, 'current_assignment.user', ''),
+      'Bureau': getResult('bureau', NO_BUREAU),
+      'Tour of Duty': getResult('post.tour_of_duty', NO_TOUR_OF_DUTY),
+      'Transfer Eligibility Date': getResult('current_assignment.estimated_end_date', NO_END_DATE),
```
Let's use our `formatDate` utility here

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -39,20 +68,20 @@ const ResultsCard = (props) => {
     {
-      'Posted': formatDate(_.get(result, 'description.date_created', NO_LAST_UPDATED_DATE), 'M.DD.YYYY'),
-      'Position Number': _.get(result, 'position_number', ''),
+      'Posted': getResult('description.date_created', NO_UPDATE_DATE),
```
`formatDate` utility

mjoyce91 commented on src/Components/ResultsCard/ResultsCard.jsx:
```diff
@@ -70,7 +99,7 @@ const ResultsCard = (props) => {
   options.compare = {
     as: 'button',
-    refKey: result.id,
+    refKey: position,
```
:+1:

mjoyce91 commented on src/Constants/SystemMessages.js:
```diff
@@ -2,24 +2,27 @@ export const DEFAULT_TEXT = 'None listed';
 export const NO_END_DATE = DEFAULT_TEXT;
 export const NO_GRADE = DEFAULT_TEXT;
 export const NO_LANGUAGES = DEFAULT_TEXT;
-export const NO_LAST_UPDATED_DATE = 'Unknown';
```
:+1:

mjoyce91 commented on src/Components/DefinitionList/DefinitionList.jsx:
```diff
@@ -30,7 +30,7 @@ const DefinitionList = (props) => {
         props.children.length ?
           props.children :
           items.map(item => (
-            <Definition id="hi" key={shortid.generate()} {...item} />
```
:+1:

burgwyn commented:
@Icemanbeta run `codeclimate analyze` and review the linting issues.  You can ignore the cognitive complexity items (I don't think this PR introduces new ones).

Icemanbeta commented:
@mjoyce91 I updated the PR with the updates I saw in the wireframe link you shared with me!
@burgwyn bwam!

<img src="https://user-images.githubusercontent.com/1786875/37861460-a6110876-2f10-11e8-8be3-accecaca4fc7.png" width="500" />

I added some commands to `package.json`:
```bash
yarn analyze
yarn analyze:eslint
```
